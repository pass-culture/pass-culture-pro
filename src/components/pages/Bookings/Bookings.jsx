import React from 'react'
import PropTypes from 'prop-types'

import DownloadButtonContainer from '../../layout/DownloadButton/DownloadButtonContainer'
import Main from '../../layout/Main'
import HeroSection from '../../layout/HeroSection/HeroSection'
import FilterByVenueContainer from './FilterByVenue/FilterByVenueContainer'
import DisplayButtonContainer from '../../layout/DisplayButton/DisplayButtonContainer'

const Bookings = ({pathToCsvFile, showButtons}) => (
  <Main name="bookings">
    <HeroSection title="Suivi des réservations">
      <p className="subtitle">
        {'Téléchargez le récapitulatif des réservations de vos offres.'}
        <br/>
        <br/>
        {'Le fichier est au format CSV, compatible avec tous les tableurs et éditeurs de texte.'}
      </p>
    </HeroSection>
    <hr/>
    <FilterByVenueContainer/>
    <div className="control flex-columns items-center flex-end">
      {showButtons && (
        <DownloadButtonContainer
          filename="reservations_pass_culture"
          href={pathToCsvFile}
          mimeType="text/csv"
        >
          {'Télécharger la liste des réservations'}
        </DownloadButtonContainer>
      )}
    </div>
    <br/>
    <div className="control flex-columns items-center flex-end">
      {showButtons && (
        <DisplayButtonContainer
          href={pathToCsvFile}
        >
          {'Afficher la liste des réservations'}
        </DisplayButtonContainer>
      )}
    </div>
  </Main>
)

Bookings.defaultProps = {
  pathToCsvFile: '',
  showButtons: false,
}

Bookings.propTypes = {
  pathToCsvFile: PropTypes.string,
  showButtons: PropTypes.bool,
}

export default Bookings
