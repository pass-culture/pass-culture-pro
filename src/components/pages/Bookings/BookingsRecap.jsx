import * as PropTypes from 'prop-types'
import React, { useCallback, useEffect, useState } from 'react'

import PageTitle from 'components/layout/PageTitle/PageTitle'
import Spinner from 'components/layout/Spinner'
import Titles from 'components/layout/Titles/Titles'
import * as pcapi from 'repository/pcapi/pcapi'

import BookingsRecapTable from './BookingsRecapTable/BookingsRecapTable'
import BookingsRecapTableLegacy from './BookingsRecapTableLegacy/BookingsRecapTableLegacy' /* eslint-disable-line react/jsx-pascal-case */
import ChoosePreFiltersMessage from './ChoosePreFiltersMessage/ChoosePreFiltersMessage'
import NoBookingsForPreFiltersMessage from './NoBookingsForPreFiltersMessage/NoBookingsForPreFiltersMessage'
import NoBookingsMessage from './NoBookingsMessage/NoBookingsMessage'
import PreFilters from './PreFilters/PreFilters'

const MAX_LOADED_PAGES = 5

const BookingsRecap = ({
  arePreFiltersEnabled,
  isUserAdmin,
  location,
  showWarningNotification,
}) => {
  const [bookingsRecap, setBookingsRecap] = useState([])
  const [isLoading, setIsLoading] = useState(arePreFiltersEnabled ? false : true)
  const [werePreFiltersApplied, setWerePreFiltersApplied] = useState(false)

  const loadBookingsRecap = useCallback(
    async preFilters => {
      setIsLoading(true)
      setWerePreFiltersApplied(true)
      setBookingsRecap([])

      // TODO(07/06/2021): To remove when 'ENABLE_BOOKINGS_PAGE_FILTERS_FIRST' feature flip has been removed
      let bookingsFilters = {}
      if (preFilters) {
        bookingsFilters = {
          venueId: preFilters.offerVenueId,
          eventDate: preFilters.offerEventDate,
        }
      }

      let currentPage = 1
      const { pages, bookings_recap: bookingsRecap } = await pcapi.loadFilteredBookingsRecap({
        page: currentPage,
        ...bookingsFilters,
      })
      setBookingsRecap(bookingsRecap)

      while (currentPage < Math.min(pages, MAX_LOADED_PAGES)) {
        currentPage += 1
        await pcapi
          .loadFilteredBookingsRecap({ page: currentPage, ...bookingsFilters })
          .then(({ bookings_recap }) =>
            setBookingsRecap(currentBookingsRecap =>
              [...currentBookingsRecap].concat(bookings_recap)
            )
          )
      }

      setIsLoading(false)
      if (currentPage === MAX_LOADED_PAGES && currentPage < pages) {
        showWarningNotification()
      }
    },
    [showWarningNotification]
  )

  useEffect(() => {
    if (!arePreFiltersEnabled) {
      loadBookingsRecap()
    }
  }, [arePreFiltersEnabled, loadBookingsRecap])

  return (
    <div className="bookings-page">
      <PageTitle title="Vos réservations" />
      <Titles title="Réservations" />
      {isUserAdmin ? (
        <NoBookingsMessage />
      ) : arePreFiltersEnabled ? (
        <>
          <PreFilters
            applyPreFilters={loadBookingsRecap}
            isLoading={isLoading}
            offerVenueId={location.state?.venueId}
          />
          {werePreFiltersApplied ? (
            bookingsRecap.length > 0 ? (
              <BookingsRecapTable
                bookingsRecap={bookingsRecap}
                isLoading={isLoading}
                locationState={location.state}
              />
            ) : isLoading ? (
              <Spinner />
            ) : (
              <NoBookingsForPreFiltersMessage />
            )
          ) : (
            <ChoosePreFiltersMessage />
          )}
        </>
      ) : bookingsRecap.length > 0 ? (
        <BookingsRecapTableLegacy
          bookingsRecap={bookingsRecap}
          isLoading={isLoading}
          locationState={location.state}
        />
      ) : isLoading ? (
        <Spinner />
      ) : (
        <NoBookingsMessage />
      )}
    </div>
  )
}

BookingsRecap.propTypes = {
  arePreFiltersEnabled: PropTypes.bool.isRequired,
  isUserAdmin: PropTypes.bool.isRequired,
  location: PropTypes.shape({
    state: PropTypes.shape({
      venueId: PropTypes.string,
      statuses: PropTypes.arrayOf(PropTypes.string),
    }),
  }).isRequired,
  showWarningNotification: PropTypes.func.isRequired,
}

export default BookingsRecap
