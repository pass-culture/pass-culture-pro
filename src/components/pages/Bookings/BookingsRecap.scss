.bookings-page {
  .br-warning {
    align-items: center;
    color: $grey-dark;
    display: flex;
    flex-direction: column;
    height: rem(400px);
    text-align: center;

    p {
      margin-top: rem(12px);
      width: rem(400px);
    }

    svg {
      height: rem(124px);
      width: rem(124px);
    }
  }

  .br-title {
    @include title3();

    display: inline-block;
    margin: 0 rem(20px) rem(24px) 0;
  }

  .reset-filters-link {
    @include body-important();

    color: $primary;
  }
}
