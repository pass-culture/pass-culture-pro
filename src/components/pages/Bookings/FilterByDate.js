import React, { PureComponent } from 'react'
import classnames from 'classnames'
import PropTypes from 'prop-types'

import { MONTH_OPTIONS, YEAR_OPTIONS } from './utils'
import {FilterByVenue} from './FilterByVenue'

export class FilterByDate extends PureComponent {

  onChangeDate = event => {
    const selectedMonth = document.getElementById("month")
    const month = selectedMonth[selectedMonth.selectedIndex].value
    const selectedYear = document.getElementById("year")
    const year = selectedYear[selectedYear.selectedIndex].value
    const updatedDate = new Date(year, month-1, 1, 12)
    this.props.selectBookingsForDate(updatedDate)
  }

  render() {
    const isDigitalChecked = false

    return (
      <React.Fragment>
        <div id="filter-by-date"
             className={classnames({'is-invisible': isDigitalChecked,})} >
          <div>
            {'Effectu√©es en :'}
          </div>
          <select
            onChange={this.onChangeDate}
            className="pc-selectbox pl24 py5 fs19"
            defaultValue="1"
            id="month"
          >
            {MONTH_OPTIONS.map(({ label, value }) => (
              <option key={value} value={value}>
                {label}
              </option>
            ))}
          </select>
          <select
            onChange={this.onChangeDate}
            className="pc-selectbox pl24 py5 fs19"
            defaultValue="2019"
            id="year"
          >
            {YEAR_OPTIONS.map(({ label, value }) => (
              <option key={value} value={value}>
                {label}
              </option>
            ))}
          </select>
        </div>
      </React.Fragment>
    )
  }
}

FilterByVenue.propTypes = {
  date: PropTypes.instanceOf(Date).isRequired,
}
