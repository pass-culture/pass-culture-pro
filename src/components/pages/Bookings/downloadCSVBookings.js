import * as pcapi from 'repository/pcapi/pcapi'
import { API_URL } from 'utils/config'

export const downLoadCSVFile = async queryParams => {
  const url = new URL(`${API_URL}/bookings/csv`)
  const params = pcapi.buildFetchParams(queryParams)
  Object.keys(params).forEach(key => url.searchParams.append(key, params[key]))

  try {
    const result = await fetch(url, { credentials: 'include' })

    if (result.status === 200) {
      const text = await result.text()
      const fakeLink = document.createElement('a')
      const blob = new Blob([text], { type: "text/csv" })
      const date = new Date().toISOString()
      fakeLink.href = URL.createObjectURL(blob)
      fakeLink.setAttribute('download', `reservations_pass_culture-${date}.csv`)
      document.body.appendChild(fakeLink)
      fakeLink.click()
      document.body.removeChild(fakeLink)

      return 'success'
    }

    return 'error'
  } catch (e) {
    return 'error'
  }
}
