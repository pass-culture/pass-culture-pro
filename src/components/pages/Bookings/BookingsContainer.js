import { connect } from 'react-redux';
import { compose } from 'redux';
import { requestData } from 'redux-saga-data'

import Bookings from './Bookings'
import { withFrenchQueryRouter, withRedirectToSigninWhenNotAuthenticated } from 'components/hocs'
import selectNonVirtualVenues from './selectNonVirtualVenues'
import selectDigitalOffers from './selectDigitalOffers'

const mapDispatchToProps = (dispatch) => ({
  loadVenues: () => {
    dispatch(requestData({
      apiPath: `/venues`,
      stateKey: 'venues',
      method: 'GET'
    }))
  },
  loadOffers: () => {
    dispatch(requestData({
      apiPath: `/offers`,
      stateKey: 'offers',
      method: 'GET'
    }))
  }
})

const mapStateToProps = state => {
  const { venues, offers } = state.data

  const allVenuesOption = {
    name: "Tous les lieux",
    id: "all",
  }

  const nonVirtualVenues = selectNonVirtualVenues(state, venues)
  const venuesOptions = [allVenuesOption, ...nonVirtualVenues]

  const allOffersOption = {
    name: "Toutes les offres",
    id: "all",
  }
  const digitalOffers = selectDigitalOffers(state, offers)
  const digitalOffersOptions = [allOffersOption, ...digitalOffers]

  return {
    venuesOptions,
    digitalOffersOptions
  }
}

export default compose(
  withRedirectToSigninWhenNotAuthenticated,
  withFrenchQueryRouter,
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
)(Bookings)

