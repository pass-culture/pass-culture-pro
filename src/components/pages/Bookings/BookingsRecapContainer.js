import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import { compose } from 'redux'

import BookingsRecap from './BookingsRecap'

export function mapStateToProps(state) {
  return {
    currentUser: state.users.currentUser,
  }
}

export default compose(withRouter, connect(mapStateToProps))(BookingsRecap)
