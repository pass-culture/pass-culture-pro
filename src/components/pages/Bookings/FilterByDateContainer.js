import {compose} from 'redux'
import {connect} from 'react-redux'
import {withFrenchQueryRouter} from 'components/hocs'

import {FilterByDate} from './FilterByDate'

export const mapDispatchToProps = (dispatch) => ({
  selectBookingsForDate: (date) => {
    dispatch(
      {
        payload: date,
        type: 'BOOKING_SUMMARY_SELECT_DATE',
      }
    )
  }
})

export const mapStateToProps = (state) => {
  return {
    date: state.bookingSummary.selectOffersSince,
  }
}

export default compose(
  withFrenchQueryRouter,
  connect(mapStateToProps, mapDispatchToProps)
)(FilterByDate)
