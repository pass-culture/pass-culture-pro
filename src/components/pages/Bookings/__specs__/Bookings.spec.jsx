import React from 'react'
import { shallow } from 'enzyme'

import Bookings from '../Bookings'
import DownloadButtonContainer from '../../../layout/DownloadButton/DownloadButtonContainer'
import DisplayButtonContainer from '../../../layout/CsvTableButton/CsvTableButtonContainer'

describe('src | components | pages | Bookings', () => {
  let props

  beforeEach(() => {
    props = {
      showButtons: false,
      pathToCsvFile: '/path/to/csv/file?with=query',
    }
  })

  it('should match snapshot', () => {
    // when
    const wrapper = shallow(<Bookings />)

    // then
    expect(wrapper).toMatchSnapshot()
  })

  describe('render', () => {
    it('should not render a download button by default', () => {
      // when
      const wrapper = shallow(<Bookings {...props} />)

      // then
      const downloadButton = wrapper.find(DownloadButtonContainer)
      expect(downloadButton).toHaveLength(0)
    })

    it('should render a download button with the right props when displayed', () => {
      // given
      props.showButtons = true

      // when
      const wrapper = shallow(<Bookings {...props} />)

      // then
      const downloadButton = wrapper.find(DownloadButtonContainer)
      expect(downloadButton).toHaveLength(1)
      expect(downloadButton.prop('href')).toBe('/path/to/csv/file?with=query')
    })

    it('should not render a display button by default', () => {
      // when
      const wrapper = shallow(<Bookings {...props} />)

      // then
      const displayButton = wrapper.find(DisplayButtonContainer)
      expect(displayButton).toHaveLength(0)
    })

    it('should render a display button with the right props when displayed', () => {
      // given
      props.showButtons = true

      // when
      const wrapper = shallow(<Bookings {...props} />)

      // then
      const displayButton = wrapper.find(DisplayButtonContainer)
      expect(displayButton).toHaveLength(1)
      expect(displayButton.prop('href')).toBe('/path/to/csv/file?with=query')
    })
  })
})
