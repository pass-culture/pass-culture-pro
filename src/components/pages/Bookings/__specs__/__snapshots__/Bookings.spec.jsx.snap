// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | pages | Bookings should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Bookings
    pathToCsvFile=""
    showButtons={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <HeroSection
          subtitle={null}
          title="Suivi des réservations"
        >
          <p
            className="subtitle"
          >
            Téléchargez le récapitulatif des réservations de vos offres.
            <br />
            <br />
            Le fichier est au format CSV, compatible avec tous les tableurs et éditeurs de texte.
          </p>
        </HeroSection>,
        <hr />,
        <withRouter(_withQueryRouter) />,
        <div
          className="control flex-columns items-center flex-end"
        />,
        <br />,
        <div
          className="control flex-columns items-center flex-end"
        />,
      ],
      "name": "bookings",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <p
            className="subtitle"
          >
            Téléchargez le récapitulatif des réservations de vos offres.
            <br />
            <br />
            Le fichier est au format CSV, compatible avec tous les tableurs et éditeurs de texte.
          </p>,
          "subtitle": null,
          "title": "Suivi des réservations",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Téléchargez le récapitulatif des réservations de vos offres.",
              <br />,
              <br />,
              "Le fichier est au format CSV, compatible avec tous les tableurs et éditeurs de texte.",
            ],
            "className": "subtitle",
          },
          "ref": null,
          "rendered": Array [
            "Téléchargez le récapitulatif des réservations de vos offres.",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
            "Le fichier est au format CSV, compatible avec tous les tableurs et éditeurs de texte.",
          ],
          "type": "p",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "hr",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": false,
          "className": "control flex-columns items-center flex-end",
        },
        "ref": null,
        "rendered": false,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": false,
          "className": "control flex-columns items-center flex-end",
        },
        "ref": null,
        "rendered": false,
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <HeroSection
            subtitle={null}
            title="Suivi des réservations"
          >
            <p
              className="subtitle"
            >
              Téléchargez le récapitulatif des réservations de vos offres.
              <br />
              <br />
              Le fichier est au format CSV, compatible avec tous les tableurs et éditeurs de texte.
            </p>
          </HeroSection>,
          <hr />,
          <withRouter(_withQueryRouter) />,
          <div
            className="control flex-columns items-center flex-end"
          />,
          <br />,
          <div
            className="control flex-columns items-center flex-end"
          />,
        ],
        "name": "bookings",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <p
              className="subtitle"
            >
              Téléchargez le récapitulatif des réservations de vos offres.
              <br />
              <br />
              Le fichier est au format CSV, compatible avec tous les tableurs et éditeurs de texte.
            </p>,
            "subtitle": null,
            "title": "Suivi des réservations",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Téléchargez le récapitulatif des réservations de vos offres.",
                <br />,
                <br />,
                "Le fichier est au format CSV, compatible avec tous les tableurs et éditeurs de texte.",
              ],
              "className": "subtitle",
            },
            "ref": null,
            "rendered": Array [
              "Téléchargez le récapitulatif des réservations de vos offres.",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": "br",
              },
              "Le fichier est au format CSV, compatible avec tous les tableurs et éditeurs de texte.",
            ],
            "type": "p",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "hr",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": false,
            "className": "control flex-columns items-center flex-end",
          },
          "ref": null,
          "rendered": false,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": false,
            "className": "control flex-columns items-center flex-end",
          },
          "ref": null,
          "rendered": false,
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
